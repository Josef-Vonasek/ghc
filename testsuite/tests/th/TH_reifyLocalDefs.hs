-- test reification of local definitions
{-# LANGUAGE CPP #-}
{-# LANGUAGE TemplateHaskell #-}
import Language.Haskell.TH.Syntax
import System.IO

-- Sidestep the staging restriction
-- printTypeOf :: String -> Q ()
#define printTypeOf(n) (addModFinalizer $ do \
                       { VarI _ t _ <- reify (mkName (n)) \
                       ; runIO $ hPutStrLn stderr (show t) \
                       })

main :: IO ()
main = print (f 1 "", g 'a' 2, h True 3)
  where
    f x y = ( x :: Int
            , let ff $(do printTypeOf("y")
                          [p| z |]
                      ) = z :: $(do printTypeOf("z")
                                    [t| () |]
                                )
               in $(do printTypeOf("x")
                       [| y :: String |]
                   )
            )
    g x y = ( $(do printTypeOf("x")
                   [| y :: Int |]
               )
            , x :: Char
            )
    h x y = ( $$(do printTypeOf("x")
                    [|| y :: Int ||]
                )
            , x :: Bool
            )
