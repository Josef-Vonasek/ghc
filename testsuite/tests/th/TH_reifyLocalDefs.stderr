ConT GHC.Types.Bool
ConT GHC.Types.Int
TupleT 0
ConT GHC.Base.String
ConT GHC.Types.Char
